shipgen/ALPACAGenerator.cxx

```diff
@@ -0,0 +1,89 @@
+#include <math.h>
+#include "TROOT.h"
+#include "TFile.h"
+#include "FairPrimaryGenerator.h"
+#include "ALPACAGenerator.h"
+#include "TDatabasePDG.h"               // for TDatabasePDG
+#include "TMath.h"                      // for Sqrt
+
+const Double_t cm = 10.; // pythia units are mm
+const Double_t c_light = 2.99792458e+10; // speed of light in cm/sec (c_light   = 2.99792458e+8 * m/s)
+
+using std::cout;
+using std::endl;
+// read events from ntuples produced
+
+// -----   Default constructor   -------------------------------------------
+ALPACAGenerator::ALPACAGenerator() {}
+// -------------------------------------------------------------------------
+// -----   Default constructor   -------------------------------------------
+Bool_t ALPACAGenerator::Init(const char* fileName) {
+  return Init(fileName, 0);
+}
+// -----   Default constructor   -------------------------------------------
+Bool_t ALPACAGenerator::Init(const char* fileName, const int firstEvent) {
+  cout << "Info ALPACAGenerator: Opening input file " << fileName << endl;
+  fInputFile  = new TFile(fileName);
+  if (fInputFile->IsZombie()) {
+    cout << "-E ALPACAGenerator: Error opening the Signal file" << fileName << endl;
+  }
+  fTree = (TTree *)fInputFile->Get("MCTrack");
+  fNevents = fTree->GetEntries();
+  fn=firstEvent;
+  fTree->SetBranchAddress("event",&event);
+  fTree->SetBranchAddress("track",&track);
+  fTree->SetBranchAddress("pdg",&pdg);              // particle pdg
+  fTree->SetBranchAddress("parent",&parent);    // parent track number??
+  fTree->SetBranchAddress("tof",&tof); // time of flight
+  fTree->SetBranchAddress("e",&e);          // incoming muon energy
+  fTree->SetBranchAddress("w",&w);                  // weight of event
+  fTree->SetBranchAddress("x",&x); // position
+  fTree->SetBranchAddress("y",&y);
+  fTree->SetBranchAddress("z",&z);
+  fTree->SetBranchAddress("px",&px);     // momentum
+  fTree->SetBranchAddress("py",&py);
+  fTree->SetBranchAddress("pz",&pz);
+  return kTRUE;
+}
+// -------------------------------------------------------------------------
+
+
+// -----   Destructor   ----------------------------------------------------
+ALPACAGenerator::~ALPACAGenerator()
+{
+ // cout << "destroy Ntuple" << endl;
+ fInputFile->Close();
+ fInputFile->Delete();
+ delete fInputFile;
+}
+// -------------------------------------------------------------------------
+
+// -----   Passing the event   ---------------------------------------------
+Bool_t ALPACAGenerator::ReadEvent(FairPrimaryGenerator* cpg)
+{
+  while (fn<fNevents) {
```

Comment:
The formatting in this while loop is very confusing! I had to read it several times to see what's in each conditional. Could you please put each statement on it's own line?
What happens when the if conditions evaluate to `false`? Are we sure the events can just be skipped?

Reply:
>The formatting in this while loop is very confusing! I had to read it several times to see what's in each conditional.

I am sorry... :/
I put some comments. I hope that it is better now.


> Could you please put each statement on it's own line?

I couldn't understand...
However, new comments would be helpful. Can you have a look at [them](https://github.com/ShipSoft/FairShip/blob/bedcc7759239cc34d391759dd076daf7fc383be8/shipgen/ALPACAGenerator.cxx)?

> What happens when the if conditions evaluate to ```false```?

In ALPACA, events are produced with an ALP and its two daughters 
OR 
Events are not produced at all due to given options -> no input file -> generator would exit.


> Are we sure the events can just be skipped?

It would be better to have a conditional just in case.
I will have a look into this.


Thanks!

Reply:
> [...]
> I put some comments. I hope that it is better now.

See my example below as an explanation of what I mean. The comments help (but it's "is add*ed*", not "is adding"). Would remove the comment on line 65.

> > Could you please put each statement on it's own line?
> 
> I couldn't understand...
> However, new comments would be helpful. Can you have a look at [them](https://github.com/ShipSoft/FairShip/blob/bedcc7759239cc34d391759dd076daf7fc383be8/shipgen/ALPACAGenerator.cxx)?
> 
> > What happens when the if conditions evaluate to `false`?
> 
> In ALPACA, events are produced with an ALP and its two daughters
> OR
> Events are not produced at all due to given options -> no input file -> generator would exit.

So if any of these if statements is not `true`, the code should exit?

Currently my understanding is that the ntuple always has groups of three entries with ALP, daughter1, daughter2. Is that correct?

> > Are we sure the events can just be skipped?
> 
> It would be better to have a conditional just in case.
> I will have a look into this.

I think you misunderstood what I meant: if these conditionals are not `true`, the track is just skipped. Is this the intended behaviour? Should there be an error if the conditional is `false`?